hack/common.sh

```diff
@@ -862,4 +865,109 @@ function os::build::find-binary() {
+function os::build::environment::create() {
+  set -o errexit
+  local golang_version="${OS_BUILD_ENV_GOLANG}"
+  local release_image="${OS_BUILD_ENV_IMAGE}"
+  local additional_context="${OS_BUILD_ENV_DOCKER_ARGS:-}"
+  if [[ -z "${additional_context}" && "${OS_BUILD_ENV_USE_DOCKER:-y}" == "y" ]]; then
+    additional_context="--privileged -v /var/run/docker.sock:/var/run/docker.sock"
+
+    if [[ "${OS_BUILD_ENV_LOCAL_DOCKER:-n}" == "y" ]]; then
+      # if OS_BUILD_ENV_LOCAL_DOCKER==y, add the local OS_ROOT as the bind mount to the working dir
+      # and set the running user to the current user
+      local workingdir
+      workingdir=$( os::build::environment::release::workingdir )
+      additional_context="${additional_context} -v ${OS_ROOT}:${workingdir} -u $(id -u)"
+    elif [[ -n "${OS_BUILD_ENV_REUSE_VOLUME:-}" ]]; then
+      # if OS_BUILD_ENV_REUSE_VOLUME is set, create a docker volume to store the working output so
+      # successive iterations can reuse shared code.
+      local workingdir
+      workingdir=$( os::build::environment::release::workingdir )
+      name="$( echo "${OS_BUILD_ENV_REUSE_VOLUME}" | tr '[:upper:]' '[:lower:]' )"
+      docker volume create --name "${name}" > /dev/null
+      additional_context="${additional_context} -v ${name}:${workingdir}"
+    fi
+  fi
+
+  # Create a new container to from the release environment
+  docker create ${additional_context} "${release_image}" "$@"
+}
+readonly -f os::build::environment::create
+
+# os::build::environment::release::workingdir calculates the working directory for the current
+# release image.
+function os::build::environment::release::workingdir() {
+  set -o errexit
+  # get working directory
+  local container
+  container="$(docker create "${release_image}")"
+  local workingdir
+  workingdir="$(docker inspect -f '{{ index . "Config" "WorkingDir" }}' "${container}")"
+  docker rm "${container}" > /dev/null
+  echo "${workingdir}"
+}
+readonly -f os::build::environment::release::workingdir
+
+# os::build::environment::cleanup stops and removes the container named in the argument
+# unless OS_BUILD_ENV_LEAVE_CONTAINER is set to a value besides 'n',
+function os::build::environment::cleanup() {
+  local container=$1
+  docker stop --time=0 "${container}" > /dev/null || true
+  if [[ -z "${OS_BUILD_ENV_LEAVE_CONTAINER:-}" ]]; then
+    docker rm "${container}" > /dev/null
+  fi
+}
+readonly -f os::build::environment::cleanup
+
+# os::build::environment::withsource starts the container provided as the first argument
+# after copying in the contents of the current Git repository at HEAD (or, if specified,
+# the ref specified in the second argument).
+function os::build::environment::withsource() {
+  local container=$1
+  local commit=${2:-HEAD}
+
+  # Generate version definitions.
+  OS_GIT_TREE_STATE=clean # set this because we will be pulling from git archive
+  os::build::get_version_vars
```

Comment:
Updated

On Wed, Jun 15, 2016 at 3:52 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> Take it up with the judge.
>
> On Wed, Jun 15, 2016 at 3:05 PM, Steve Kuznetsov <notifications@github.com
> > wrote:
>
>> In hack/common.sh
>> <https://github.com/openshift/origin/pull/9278#discussion_r67225534>:
>>
>> > +  if [[ -z "${OS_BUILD_ENV_LEAVE_CONTAINER:-}" ]]; then
>> > +    docker rm "${container}" > /dev/null
>> > +  fi
>> > +}
>> > +readonly -f os::build::environment::cleanup
>> > +
>> > +# os::build::environment::withsource starts the container provided as the first argument
>> > +# after copying in the contents of the current Git repository at HEAD (or, if specified,
>> > +# the ref specified in the second argument).
>> > +function os::build::environment::withsource() {
>> > +  local container=$1
>> > +  local commit=${2:-HEAD}
>> > +
>> > +  # Generate version definitions.
>> > +  OS_GIT_TREE_STATE=clean # set this because we will be pulling from git archive
>> > +  os::build::get_version_vars
>>
>> not a huge fan of this type of function that spits out envars instead of
>> returns
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/openshift/origin/pull/9278/files/8437cd96443c7aee0b866981c9a57910926e0646#r67225534>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe/ABG_p8fDNpoyGRvQWzgaDAz8r9KtAJWTks5qMEz1gaJpZM4IzMty>
>> .
>>
>
>